l1-contracts/contracts/bridge/L1ERC20Bridge.sol

```diff
@@ -6,122 +6,303 @@ import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+    function startErc20BridgeInitOnChain(
+        uint256 _chainId,
+        uint256 _mintValue,
+        bytes[] calldata _factoryDeps,
+        uint256 _deployBridgeImplementationFee,
+        uint256 _deployBridgeProxyFee
+    ) external payable {
+        {
+            require(l2BridgeAddress[_chainId] == address(0), "L1EB: bridge already deployed");
+            // We are expecting to see the exact three bytecodes that are needed to initialize the bridge
+            require(_factoryDeps.length == 3, "L1EB: invalid number of factory deps");
+            require(factoryDepsHash == keccak256(abi.encode(_factoryDeps)), "L1EB: invalid factory deps");
+        }
+        bool ethIsBaseToken = bridgehub.baseToken(_chainId) == ETH_TOKEN_ADDRESS;
         {
-            // Data to be used in delegate call to initialize the proxy
-            bytes memory proxyInitializationParams = abi.encodeCall(
-                IL2ERC20Bridge.initialize,
-                (address(this), l2TokenProxyBytecodeHash, _governor)
+            if (ethIsBaseToken) {
+                _mintValue = msg.value;
+            } else {
+                require(msg.value == 0, "L1EB: msg.value > 0, base token");
+            }
+            // The caller miscalculated deploy transactions fees
+            require(_mintValue == _deployBridgeImplementationFee + _deployBridgeProxyFee, "L1EB: invalid fee");
+        }
+        bytes32 l2BridgeImplementationBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[0]);
+        bytes32 l2BridgeProxyBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[1]);
+        {
+            // Deploy L2 bridge implementation contract
+            (address bridgeImplementationAddr, bytes32 bridgeImplTxHash) = BridgeInitializationHelper
+                .requestDeployTransaction(
+                    ethIsBaseToken,
+                    _chainId,
+                    bridgehub,
+                    _deployBridgeImplementationFee,
+                    l2BridgeImplementationBytecodeHash,
+                    "", // Empty constructor data
+                    _factoryDeps // All factory deps are needed for L2 bridge
+                );
+
+            // Prepare the proxy constructor data
+            bytes memory l2BridgeProxyConstructorData;
+            {
+                address proxyAdmin = readProxyAdmin();
+                address l2ProxyAdmin = AddressAliasHelper.applyL1ToL2Alias(proxyAdmin);
+                address l2Governor = AddressAliasHelper.applyL1ToL2Alias(owner());
+                // Data to be used in delegate call to initialize the proxy
+                bytes memory proxyInitializationParams = abi.encodeCall(
+                    IL2ERC20Bridge.initialize,
+                    (address(this), l2TokenProxyBytecodeHash, l2ProxyAdmin, l2Governor)
+                );
+                l2BridgeProxyConstructorData = abi.encode(
+                    bridgeImplementationAddr,
+                    l2ProxyAdmin,
```

Comment:
Ah, that is bad. Ok, so we cannot have governor functions on the l2 bridges, because it has to be the governor that upgrades them. Ok, I don't think we need governor priveleges now, but we might need it later for some things. Also other bridges (Lido), will definitely need things like that. 

Probably having an L2Governor contract that can only be called from L1Governor would work. It is not ideal, but i can't think of anything better for these contracts. 

The real solution would be to use UUPS, because those don't have this issue.